plugin/template.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"go/parser"
+	"go/printer"
+	"go/token"
+	"sort"
+	"text/template"
+
+	"github.com/thriftrw/thriftrw-go/internal/goast"
+	"github.com/thriftrw/thriftrw-go/plugin/api"
+
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+// TemplateOption provides optional arguments to GoFileFromTemplate.
+type TemplateOption struct {
+	apply func(*goFileGenerator)
+}
+
+// TemplateFunc is a TemplateOption that makes a function available in the
+// template.
+//
+// The function may be anything accepted by text/template.
+//
+// 	GoFileFromTemplate(
+// 		filename,
+// 		`package <lower "HELLO">`,
+// 		TemplateFunc("lower", strings.ToLower),
+// 	)
+func TemplateFunc(name string, f interface{}) TemplateOption {
+	return TemplateOption{apply: func(t *goFileGenerator) {
+		t.funcs[name] = f
+	}}
+}
+
+// GoFileImportPath is a TemplateOption that specifies the intended import path
+// for the file being generated.
+//
+// If specified, this changes the behavior of the `formatType` template function
+// to NOT import this package and instead use the types directly since they are
+// available in the same package.
+func GoFileImportPath(path string) TemplateOption {
+	return TemplateOption{apply: func(t *goFileGenerator) {
+		t.importPath = path
+	}}
+}
+
+// goFileGenerator generates a single Go file.
+type goFileGenerator struct {
+	importPath string
+	funcs      map[string]interface{}
```

Comment:
can you specify that this is template functions, since it's in a struct with `globals` and `imports`, it's easy to mistake this as a list of functions in the file